### DISCUSSION TITLE
Use serverless-api-gateway-throttling plugin

### DISCUSSION THREAD

**Discussion by ACannuniRP** [brefphp/bref#918] [Support] [answered]
Hello Community,
it is the first time exploring Bref and it is amazing to see how easy is to deploy an application. I'm pretty concerned about the security of the application (that at the moment is only on dev environment) and I would like to use the Serverless Framework plugin called [serverless-api-gateway-throttling](https://www.serverless.com/plugins/serverless-api-gateway-throttling) for setting the throttling on all the endpoints (at the moment only one).
As I found in the instructions I added the plugin in the plugins section of the YAML file:
```
...
plugins:
    # We need to include the Bref plugin
    - ./vendor/bref/bref
    # In order to avoid expensive bills
    - serverless-api-gateway-throttling
```

and a custom property just below:
```
custom:
    # Configures throttling settings for the API Gateway stage
    # They apply to all http endpoints, unless specifically overridden
    apiGatewayThrottling:
        maxRequestsPerSecond: 10
        maxConcurrentRequests: 5
...
```

However when deploying I receive the following error message:
```
...
Serverless: Validating template...
 
 Error ---------------------------------------------------
 
  Error: The CloudFormation template is invalid: Unresolved resource dependencies [ApiGatewayRestApi] in the Outputs block of the template
      at /usr/local/lib/node_modules/serverless/lib/plugins/aws/deploy/lib/validateTemplate.js:20:13
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              14.8.0
     Framework Version:         2.36.0
     Plugin Version:            4.5.3
     SDK Version:               4.2.2
     Components Version:        3.8.3
```

I understand that the problem is that "ApiGatewayRestApi" is not visible (if I use the plugin) to Serverless Framework deployment but I can still see this section in the file `.serverless/cloudformation-template-update-stack.json`:
```
    "RestApiIdForApigThrottling": {
      "Description": "Rest API Id",
      "Value": {
        "Ref": "ApiGatewayRestApi"
      }
    }
```

I read the instructions here but I found them very "brief" and they didn't help me.
Could you help me in understanding how to use a plugin in Bref?
Thank you in advance.

**Comment by mnapoli** [comment, marked as answer]
Hey, it seems to be blocked by this: https://github.com/DianaIonita/serverless-api-gateway-throttling/issues/7

You would have to use API Gateway v1 to use the plugin it seems.

**Reply by ACannuniRP** [reply]
Hi @mnapoli , thank you very much, it is definitively my fault for not managing to find the issue. Thanks for writing Bref!

**Reply by mnapoli** [reply]
Hey no problem, good luck with the plugin :)

**Reply by ACannuniRP** [reply]
Just a quick update that may result helpful to anyone reading this post: I just tested another plugin called [serverless-apigateway-route-settings](https://www.serverless.com/plugins/serverless-apigateway-route-settings) and I achieved the correct result managing in setting the default stage throttling.
